servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java

```diff
@@ -0,0 +1,107 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.ConnectionFactory;
+import io.servicetalk.client.api.LoadBalancedConnection;
+import io.servicetalk.client.api.LoadBalancer;
+import io.servicetalk.client.api.LoadBalancerFactory;
+import io.servicetalk.client.api.ServiceDiscovererEvent;
+import io.servicetalk.concurrent.api.Publisher;
+
+import java.util.function.Predicate;
+
+/**
+ * {@link LoadBalancerFactory} that creates {@link LoadBalancer} instances which use a round robin strategy
+ * for selecting addresses. The created instances have the following behaviour:
+ * <ul>
+ * <li>Round robining is done at address level.</li>
+ * <li>Connections are created lazily, without any concurrency control on their creation.
+ * This can lead to over-provisioning connections when dealing with a requests surge.</li>
+ * <li>Existing connections are reused unless a selector passed to {@link LoadBalancer#selectConnection(Predicate)}
+ * suggests otherwise. This can lead to situations where connections will be used to their maximum capacity
+ * (for example in the context of pipelining) before new connections are created.</li>
+ * <li>Closed connections are automatically pruned.</li>
+ * <li>If {@link #eagerConnectionShutdown} is set to {@code true}, connections are immediately closed
+ * for an {@link ServiceDiscovererEvent#isAvailable() unavailable} address. When {@code false} is used, connections
+ * to addresses marked as {@link ServiceDiscovererEvent#isAvailable() unavailable} are used for requests,
+ * but no new connections are created for them. In case the address' connections are busy, another host is tried.
+ * If all hosts are busy, selection fails with a {@link io.servicetalk.client.api.ConnectionRejectedException}.</li>
+ * </ul>
+ *
+ * @param <ResolvedAddress> The resolved address type.
+ * @param <C> The type of connection.
+ */
+public final class RoundRobinLoadBalancerFactory<ResolvedAddress, C extends LoadBalancedConnection>
+        implements LoadBalancerFactory<ResolvedAddress, C> {
+
+    static final boolean EAGER_CONNECTION_SHUTDOWN_ENABLED = true;
+    private final boolean eagerConnectionShutdown;
+
+    private RoundRobinLoadBalancerFactory(boolean eagerConnectionShutdown) {
+        this.eagerConnectionShutdown = eagerConnectionShutdown;
+    }
+
+    @Override
+    public <T extends C> LoadBalancer<T> newLoadBalancer(
```

Comment:
do we need a factory and a builder? is just a builder sufficient? if just a builder is sufficient we could have it be an inner type of `RoundRobinLoadBalancer` and avoid deprecation of the existing type.

Reply:
All our API at the client level target `LoadBalancerFactory`. It's hard to operate with the `LoadBalancer` directly because you have to coordinate events publisher and connection factory. So, `LoadBalancerFactory` is a helpful abstraction.
Here we have a builder for the `LoadBalancerFactory` because we want to set some configuration options (we plan to add more in a follow-up PR). Users can use the result of this builder to create as many clients as necessary, reusing a single `LoadBalancerFactory`.